src/plugins/intel_gpu/src/graph/graph_optimizer/set_required_layouts.cpp

```diff
@@ -0,0 +1,55 @@
+// Copyright (C) 2018-2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+
+#include "pass_manager.h"
+#include "data_inst.h"
+#include "mutable_data_inst.h"
+#include "program_node.h"
+#include "intel_gpu/runtime/engine.hpp"
+#include "runtime/cldnn_itt.hpp"
+#include <iostream>
+#include "to_string_utils.h"
+#include "intel_gpu/runtime/debug_configuration.hpp"
+#ifdef ENABLE_ONEDNN_FOR_GPU
+#include <oneapi/dnnl/dnnl.hpp>
+#include "impls/onednn/utils.hpp"
+#include "impls/onednn/convolution_onednn.hpp"
+#endif
+
+using namespace cldnn;
+
+void set_required_layouts::run(program& p) {
+    OV_ITT_SCOPED_TASK(itt::domains::CLDNN, "CLDNN::pass::SetRequiredLayouts");
+
+    auto& engine = p.get_engine();
+    const auto& device_info = engine.get_device_info();
+
+    if (!device_info.supports_immad)
+        return;
+
+#ifdef ENABLE_ONEDNN_FOR_GPU
+    for (auto n : p.get_processing_order()) {
+        if (!n->is_type<convolution>()
```

Comment:
I think this pass should be generic as we'll need something similar for all primitive types. So maybe this layout selection logic should now be a part of program_node and then flow would look like:
```cpp
apply_pass<select_preferred_formats>();
...
apply_pass<optimize_formats>();
...
apply_pass<insert_reorders>();
```

And `select_preferred_formats` pass would do something simple like:
```cpp
for (auto n : p.get_processing_order()) {
    n.select_preferred_formats();
}
```

Initial solution may assign format::any by default and be overridden for convolution node only with onednn as preferred impl type.

Reply:
We now have functions like `get_preferred_impl_type` in layout_optimizer.cpp. But you want to place it in program_node.cpp?

Reply:
Hmm, probably it should be still a part of layout optimizer. 
```cpp
for (auto n : p.get_processing_order()) {
    auto formats = lo.get_preferred_formats(n); // need to decide on return types, but probably here should be a single method which returns both input and output layouts.
    n.set_preferred_formats(formats);
}
```